examples/link-viewer-cameras.py

```diff
@@ -0,0 +1,53 @@
+"""
+Link cameras between two viewers
+================================
+
+This lets you zoom in on different layers side by side instead of overlaying
+them. [1]_
+
+.. [1] https://forum.image.sc/t/zooming-in-on-the-same-region-on-two-different-layers/95659/5
+
+.. tags:: gui, layers, visualization-advanced, multicanvas
+"""
+from skimage import data, filters
+
+import napari
+
+# example data
+coins = data.coins()
+binary = coins > filters.threshold_otsu(coins)
+
+# make two viewers
+viewer0 = napari.Viewer()
+viewer1 = napari.Viewer()
+
+# add a layer to each
+viewer0.add_image(coins)
+viewer1.add_labels(binary)
+
+# note: check available events with:
+# viewer0.camera.events.emitters
+
+# reusable functions to set camera zoom and center
+def set_zoom(camera, value):
+    camera.zoom = value
+
+def set_center(camera, value):
+    camera.center = value
+
+# hook everything up
+viewer0.camera.events.zoom.connect(
+    lambda ev: set_zoom(viewer1.camera, ev.value)
+)
+viewer1.camera.events.zoom.connect(
+    lambda ev: set_zoom(viewer0.camera, ev.value)
+)
+viewer0.camera.events.center.connect(
+    lambda ev: set_center(viewer1.camera, ev.value)
+)
+viewer1.camera.events.center.connect(
+    lambda ev: set_center(viewer0.camera, ev.value)
+)
```

Comment:
We should not use lambdas in examples in signal connection. It is a bad pattern when someone tries to do this in plugin code.

Reply:
(Would a tz.curried function be better?)

Reply:
Because lambdas keep hard references to objects from their namespaces, that leads to leaking objects and preventing gc from deleting obsolete objects. And when we reffering Qt objects, it may refer to objects that were deleted by qt and leads to segfaults. 

Of course, it is not this case, but someone may inspire by example and cause problems. 

I suggest use `event.source`

Reply:
Could we put that info as a comment or in the header of the example? "If you want to do this in a plugin be aware x y z"
I like how straightforward this is. Not intimidating and easy to understand.

Reply:
Maybe. I would also like to point this discussion (similar) https://github.com/pyapp-kit/app-model/pull/183 and We should check our plugin create docs.

Reply:
> I suggest use `event.source`

Can you provide an example of what you mean by this?

Reply:
Ok so reading through pyapp-kit/app-model#183, the idea is that *in Qt-land*, if you have a signal and you connect it to a method, PyQt will check whether the object having that method exists before triggering the callback. (Right?)

However, for VisPy Events, I don't think the same is true? ie VisPy will call a callback regardless of whether it's a method or a lambda or a partial? And in this case we are *not* actually calling Qt classes, so the problem is relegated downstream? (ie I think the `viewer.camera` connections in the second viewer should be checking.)

In general, I think we should think long and hard about how to make expressions like the above, ie *perfectly valid Python*, be acceptable practice for our users. This should all be implementation details, not API restrictions.

Reply:
> However, for VisPy Events, I don't think the same is true? ie VisPy will call a callback regardless of whether it's a method or a lambda or a partial?

Qt also allows connecting lambdas. psygnal allows connecting lambdas, vispy and napari.events allow connecting lambdas. Each time someone uses lambda, it may lead to keeping a hard reference to objects that should be removed. So it leads to memory leaking. If somewhere in references, there are Qt objects it may lead to segfault. 

>  And in this case we are _not_ actually calling Qt classes, so the problem is relegated downstream? (ie I think the `viewer.camera` connections in the second viewer should be checking.)

But we connect Qt objects to camera events. So this lambda keeps reference to `viewer0`

```python
viewer1.camera.events.center.connect(
    lambda ev: set_center(viewer0.camera, ev.value)
)
```

and prevents for cleanup of `viewer0` until the `viewer1` is alive. And even after deleting the last reference to `viewer0` and `viewer1` we have a circular dependency that needs to be detected in `gc` and will be removed in unpredicted time.   

> In general, I think we should think long and hard about how to make expressions like the above, ie _perfectly valid Python_, be acceptable practice for our users. This should all be implementation details, not API restrictions.

Lambdas are bad. If we need a one-liner, it should be partial. But best is to manage references using objects and methods (ie. user should connect method of widget instead of partial in plugin). 


If you think that we should better support lambdas it is possible, but it will lead to multiple hour development of inspection of ripping lambdas on connection and storing weak references to napari objects.

Reply:
> If we need a one-liner, it should be partial.

how do partials help?

> But best is to manage references using objects and methods (ie. user should connect method of widget instead of partial in plugin).

I don't want to change our entire architecture to get/set instead of attrs/properties because of this. If what we want is to avoid strong references, maybe we can provide our own partial that wraps the input arguments in weakrefs.

```python
from napari.utils import partial

callback = partial(set_zoom_from_event, viewer0)
```

then partial keeps a weakref to viewer0 instead of a strong ref. (Or does partial already do this?)

I presume not *every* object should be a partial — in the discussion in pyapp-kit/app-model#183, you talk about "simple" and "complex" objects but I don't know what this distinction means... Is "simple" just a built-in type?

Reply:
> how do partials help?\n\n

In this case lambda is storing reference to viewer? And partial will store reference only to camera.

Reply:
I think that you should update this example to use this utility https://github.com/napari/napari/blob/main/napari%2Futils%2Fevents%2Fevent_utils.py#L63

(Commenting from mobile so respond in parts).

Reply:
> I presume not every object should be a partial — in the discussion in pyapp-kit/app-model#183, you talk about \"simple\" and \"complex\" objects but I don't know what this distinction means... Is \"simple\" just a built-in type?

By simple, I mean object, which leaking is not a problem. So number, text, aray of texts, simple mapping dictionary, list of ints, pydantic model ith simple fields are simple. But Layer, Viewer, non small numpy array, some evented object, list of layers, dict with mapping from name to viewer etc, means complex in this case

Reply:
> By simple, I mean object, which leaking is not a problem.

That still isn't clear to me. And I think it's a critical question because, for example, if I make a partial with a reference to particular number and I connect it, I want to keep that reference so the parameter, which is needed at call time, is not gc'd. Can one write a function `is_simple_object` that determines this?

Reply:
> Can we make self-disconnecting connections using weakrefs? 🤔

(I see that this is what the utility does. *But* it does this on callback time, rather than when the object is destroyed using the weakref callback, and I'm also wondering whether we can do this with `connect` directly, rather than with a utility. Or maybe with a decorator?)

Reply:
> That's interesting. Can we make self-disconnecting connections using weakrefs? 🤔

we have tried in the past, and we ended with segfaults.. (Qt related..). Whis is why there are lines:

```
        if (ob := ref()) is None:
            emitter.disconnect(_cb)
            return
```

in the code. 

> Can one write a function `is_simple_object` that determines this?

Such a function will need to go through the whole dependency graph, and this could be really time-consuming. And still may be a problem with edge cases. 

But we could do simpler thing that search in depth 3 or 5 only and validate if some well known problematic class is present (layer, viewer, np.ndarray of size above 1000 etc).

Reply:
@jni do you still plan to change the lambdas before merging?

Reply:
Yes, I think @Czaki will yell at me otherwise. 😜 Not sure whether I'll make it before 0.5.2, but looking at the discussion, I think I could do it with the partial + some documentation. Then we can make an issue for our fancy weak partial utilities.